sphinx/ext/intersphinx/_load.py

```diff
@@ -21,55 +22,96 @@
 
     from sphinx.application import Sphinx
     from sphinx.config import Config
-    from sphinx.ext.intersphinx._shared import InventoryCacheEntry
+    from sphinx.ext.intersphinx._shared import (
+        IntersphinxMapping,
+        InventoryCacheEntry,
+        InventoryLocation,
+        InventoryName,
+        InventoryURI,
+    )
     from sphinx.util.typing import Inventory
 
 
 def normalize_intersphinx_mapping(app: Sphinx, config: Config) -> None:
-    for key, value in config.intersphinx_mapping.copy().items():
+    # URIs should NOT be duplicated, otherwise different builds may use
+    # different project names (and thus, the build are no more reproducible)
+    # depending on which one is inserted last in the cache.
+    seen: dict[InventoryURI, InventoryName] = {}
+
+    errors = []
+    for name, value in config.intersphinx_mapping.copy().items():
+        # ensure that intersphinx projects are always named
+        if not isinstance(name, str):
+            msg = __('project identifier must be a string')
+            errors.append((msg, name, name))
```

Comment:
Nitpicking I think, but: in these cases where we supply `name` as the third argument (`value`, logically), the resulting error messages could be a bit confusing.

Reply:
Errr... I didn't get this one. Could you explain with an example please?

Reply:
> Nitpicking I think, but: in these cases where we supply `name` as the third argument (`value`, logically), the resulting error messages could be a bit confusing.

I agree. See [`8ada5cd` (#12083)](https://github.com/sphinx-doc/sphinx/pull/12083/commits/8ada5cd299b29208a149ef9372e7963b29b2d3b8) for an improvement...